blit/doc/joff.1
.TH JOFF 1 Blit
.SH NAME
joff \- symbolic debugger
.SH SYNOPSIS
.B joff
[-\fBo\fIfile\fP]
.SH DESCRIPTION
.I Joff
is a source-level symbolic debugger for C programs running under
.I mpx.
Running asynchronously and
communicating by means of the keyboard and mouse in its
own layer, it provides for controlled execution of another process running under
.I mpx.
It takes advantage of the
source symbol tables generated by 
.I mcc's
\-g option.
.SH MODEL
Most commands manipulate and are interpreted with respect to three
major values:
.IP
i) \fIstate\fP: the active statement in the debugged process
.IP
ii) \fIcurrent context\fP: the active statement in one of the functions on the call stack
.IP
iii) \fIcurrent location\fP: an arbitrary address within Blit memory
.LP
If there is a state, it is displayed in a status line at the top of
.I joff's
layer; otherwise an appropriate message appears there.
If there is a current context, the last item on the button 3 menu
is ``\fIfunc_id\fP() vars'', where \fIfunc_id\fP is the function.
If the current location is non-zero, it appears as the normal command prompt. 
.SH INTERACTION
Apart from the status line (which video inverts momentarily whenever
its contents change),
.I joff's
layer scrolls conventionally.
.PP
.I Joff
usually prompts ``:'' (or ``\fIaddress\fP:'' with the current location)
when it is ready to accept a command entered from either the keyboard or
a mouse menu.
If a ``dead mouse'' icon replaces the cursor, the mouse may not be used.
This happens only when prompting for the symbol table file and
when waiting for the ``Debug'' selection from
.I mpx's
main menu.
The ``menu'' icon that replaces the cursor shows when there are menus available
on buttons 2 and/or 3.
It is always safe to raise a menu. If it is not what you want, don't select
anything and there is no harm done.
When
.I joff
is busy for a long period,
a ``coffee cup'' icon replaces the cursor.
.PP
If invoked with \-o,
.I joff
examines a program directly from its object file.
In this mode
.I joff
need not run under
.I mpx.
.SH MOUSE
Buttons 2 and 3 display command menus.
Button 3 commands treat the process only at the source language level.
Button 2 commands allow further access to the 68000 as raw memory.
Selecting nothing from a menu is always safe; it has no effect.
Selecting ``more...'' at the top or bottom of a menu, moves to another portion of
a menu which is too large to be displayed.
Some commands call for confirmation by a second button hit on a ``skull'' cursor.
.SH Button 3
.TP
.B layer
Selects a new layer for examination.
The layer is identified by picking
.B Debug
from the
.I mpx
menu and then hitting the layer with the ``bullseye'' cursor.
The layer is flashed by video invert.
There is then a prompt for an object file with \-g symbol tables.
Response is a file name, or null (carriage return) to select the default as shown,
or ``\-'' to proceed without symbols.
The default file name is taken from the
.I 68ld
command which loaded the layer.
It is therefore best to give an
.I absolute
pathname when using
.I 68ld
and
.I jx
in shell scripts and programs.
.TP
.B quit
Must be confirmed.
.TP
.B breakpts
Repeatedly offers a further menu on button 3 with the commands
.B list [\fIn\fB], clear all
and a list of functions in the program.
The value in brackets with
.B list
shows how many breakpoints are currently set.
The commands list and clear them all, respectively.
.B Clear all
must be confirmed.
Selecting a function identifier offers a terminal menu with
.ce
.B call
.ce
.B return
.ce
.B both
.ce
.B none
One of these is tagged by ``>'', indicating what breakpoints are set.
Selecting one modifies the breakpoints on the function. 
The breakpoint at function call is
.I after
the stack frame has been advanced,
but
.I before
initialization of the registers and automatics.
The breakpoint on function return is
.I after
the return value has been set, but
.I before
the stack frame is retracted.
It is safe to set and clear breakpoints
.I while the process is running.
To set breakpoints before the program starts, use the \-z option
on
.I 68ld
to leave the process halted after loading.
Breakpoints can only be set on individual C statements by means
of the keyboard (below).
.TP
.B globals
Repeatedly offers a menu with the program's globals.
Each one selected is displayed.
.TP
.B stmt step
Executes a single statement of the program.
After it is executed the status line shows the next statement to be
executed.
See the .sd keyboard command for specifying directories to search for
source files.
If the statement returns to the calling function,
the current context will not set and
\fBfunction\fP or \fBtraceback\fP must be used explicitly before
local variables are accessible.
Statement stepping can only be used after stopping on breakpoint.
It isn't available when the process is merely suspended by \fBhalt\fP.
.TP
.B go
Resumes execution of the process.
.TP
.B halt
Suspends execution of the process.
.TP
.B traceback
Lists the active function stack frames from the one that
is executing back to main().
Each line in the traceback has the form
.ce
\fIfile\fP:\fIline\fP[+\fIoffset\fP] in \fIfunction(arg, ... arg)\fP
giving the statement being executed as source file, line number
(and instruction byte offset) within a function and its arguments.
The deepest stack frame becomes the current context.
.TP
.B function
Offers a menu of functions (in traceback order) from which one may be selected
as current context.
.TP
.I func_id\fB() vars
Repeatedly offers a menu of the local variables, arguments and statics 
which are alive at the active statement in the current stack frame.
Each variable is classified as ``lcl'' (local), ``reg'' (register), ``sta''
(static) or ``arg'' (argument).
.SH Pointer Chasing
.PP
If a variable selected from the global or local menu is a structure,
a pointer to a structure or an array, it is displayed as one of
.ce 3
\fIvariable\fP .  <field>
\fIvariable\fP \-> <field>
\fIvariable\fP    <index>
.LP
respectively.
The menu on button 3 can then be used to complete the expression and
display its value.
While the expression this yields is still an aggregate,
a futher selection must be made to extend it.
Making no selection or hitting button 1 or 2 ``backs up'' out of an 
expression.
.PP
If the structure reached is a Point, Rectangle, Texture, Bitmap or Layer
a special operation point(), outline(), texture() or bitblt()
also appears in the menu to display the object graphically.
.SH Button 2
.PP
Button 2 has commands which root around in memory displaying it as bytes, words
(16 bits), long words (32 bits), structures, unions and enumerations.
Before using this, the command
.ce 1
 . \fIexpression\fP
is needed to set the current location to the value of the expression.
.PP
The model for this button is that the menu itself is a window into memory.
Selections from the menu change the location, size and format of the objects
seen the next time it is raised and can
copy useful ones into the scrolling area.
.PP 
The top three menu items are the location and
value for the current memory location (tagged by ``.'')
and its two neighbors.
Selecting one of these three writes it into the scrolling
area and makes that location the current one.
.PP
Selecting \fBbyte, word, long\fP fixes the size of objects viewed.
Selecting \fBdecimal, octal, hex, ascii\fP fixes the radix.
The ascii radix displays each byte as a C character
constant (e.g. 'a', '\\n', '\\277').
.PP
For pointer chasing, if the object size is \fBlong\fP and
the value at the current location makes sense as an address,
the location referenced and its value are shown (tagged by ``*''.)
It can be picked to become the current location.
.PP
\fBString\fP shows the zero-terminated strings starting at the current location,
and at current\-4 and current+4.
Long strings are truncated.
.PP
\fBstruct\f and \fBenum\fP present menus of the structures and enumerations
in the program.
Selecting one interprets memory as a vector of such objects.
Most structures are horribly truncated in the menu, but are displayed in full
when selected and displayed in the scrolling area.
.PP
If a there is a symbolic form for the current location it is displayed as
an item tagged by ``.''.
Selecting this gives a fuller symbolic form.
.PP
A final item may appear in this menu.
If
.I joff
is unable to display an object when given its symbolic name it saves
the address and places it in the menu.
It may be picked to become the current location.
.SH KEYBOARD
.I joff
repeatedly prompts with the address of the
current memory location and accepts one or more of the
following commands separated by semi-colons.
Constants may be unsigned decimal, octal or
hexadecimal, in C notation.
.TP
.B .la
Select a new layer to debug.
.TP
.B .q
Quit
.TP
.B \fIexpression\fP
Evaluate and print its value.
The following operators are supported:
.ce 2
+ \- * / %  \->  [ ] . = (binary)
& * \-            (unary)
Functions may also be called, but are executed by
.I joff's
process (while the user's is suspended if necessary.)
Structures may be expressed as \fIstruct_id(expr, ... expr)\fP,
e.g. Point(100,200).
Identifiers are bound to variables as they they would be in the current
context using C's scope rules.
.TP
.B .v
Print all local variables visible from the current context.
.TP
.B .t
Display a function traceback. Set the current context
to the deepest user function. 
.TP
.B .h
Halt the process.
.TP
.B .g
Go the process.
.TP
.B .bc \fIfunction_id\fP
Set a breakpoint at function call.
.TP
.B .br \fIfunction_id\fP
Set a breakpoint at function return.
.TP
.B .bs \fIfile line_no\fP
Set a breakpoint at a given line within a source file.
.TP
.B .cc \fIfunction_id\fP
Clear breakpoint at function call.
.TP
.B .cr \fIfunction_id\fP
Clear breakpoint at function return.
.TP
.B .cs \fIfile line_no\fP
Clear breakpoint at a given line within a source file.
.TP
.B .bl
List all breakpoints.
.TP
.B .ca
Clear all breakpoints.
.TP
.B .c [\fIcount\fP]
Set the new current context to the function which the present one calls.
When given, iterate the number of times specified by the constant.
.TP
.B .r [\fIcount\fP]
Set the new current context to the the function to which the present one returns.
When given, iterate the number of times specified by the constant.
.TP
.B .  \fIexpression\fP
Set the address of current memory location to the value of the expression.
.TP
.if n .B [+|-] \fI[offset\fP]
.if t .B [\(+-] [\fIoffset\fP]
Set the current location relative to itself by the given constant times
the size of the object last displayed.
The null command (a carriage return) is equivalent to +1.
.TP
.B .x
Find a symbolic name for the current memory location.
.TP
.B .b
Display the byte at the current location.
.TP
.B .w
Display the word (16 bits) at the current location.
.TP
.B .l 
Display the long word (32 bits) at the current location.
.TP
.B .i
Display the instruction at the current location.
.TP
.B \fIstructure_id\fP
The name of a defined structure
displays memory from the current location as an instance the structure.
Also unions and enumerations.
.TP
.B %\fIregister\fP
Set the current location to where the specified register
in the current context was saved and display as a long (32 bits).
\fIregister\fP is [ad][0\-7], sp, fp.
.TP
.B \(**
Set the address of the current location to the 32 bit address
at the current location.
.TP
.B .li \fIfile line_no\fP
Display the first instruction of code generated for
the statement beginning at a line number of a source file.
.TP
.B .= \fIexpression\fP
Write value of expression to the current memory location.
Operand length is that of the
previous .b, .w or .l command (or button 2 equivalent).
.TP
.B < \fIfile\fP
Read command script from a file.
.TP
.B .bi
Set breakpoint on the instruction at the current location.
.TP
.B .ci
Clear breakpoint at the current location.
.TP
.B .si [\fIcount\fP]
Single step the next instruction or number of instructions.
(At present this only works if already stopped on a breakpoint.)
.TP
.B .ss [\fIcount\fP]
Single step the next C statement or number of statements.
(At present this only works if already stopped on a breakpoint.)
.TP
.B .\fIbase\fP
Set radix for display of addresses, where \fIbase\fP is 8, 10 or 16.
.TP
.B .pwd
Print working directory.
.TP
.B .cd \fIdirectory\fP
Change directory
.TP
.B .sd \fIdirectory\fP
Add the directory to the front of the list of directories to be
searched for source files for displaying source lines during
statement stepping.
.SH "SEE ALSO"
mcc(1) sdb(1) mpx(1) 68ld(1)
.SH DIAGNOSTICS
The diagnostics produced by
.I joff
are intended to be self-explanatory.
Messages prefaced by ``68000: '' are generated and printed autonomously by
the terminal program when the state of the debugged process is not as expected.
They appear in exceptional situations where it is hard for
.I joff
to keep track of events.
Examples are: manually patching
breakpoint traps, wild stores to code or control memory, reloading of the 
debugged layer.
Once such a diagnostic is given there are no guarantees.
.SH BUGS
.I joff
is still evolving and has some rough edges.
The .= command can write anywhere in memory.
Anonymous structures, unions and enumerations are not supported.
For example:
.IP
typedef { int field } Record;
.LP
introduces an anonymous structure.
A simple remedy is to name the structure thus:
.IP
typedef Record { int field } Record;
.PP


blit/src/joff/host/joff.c
#include "common.h"
#include "user.h"
#include "frame.h"
#include "../traps.h"

static delta = 0;

char *unixline( str )
char *str;
{
	int i;
	char c;

	for( i = 0;; ){
		c = getchar();
		if( c == '\r' || c == '\n' ) break;
		if( c != '\b' ) printf( "%c", str[i++] = c );
		else if( i > 0 ) printf( "\b%c\b", str[--i] );
		if( c == 004 ) break;
	}
	str[i] = '\0';
	return str;
}

char *readuser( str )
char *str;
{
	char prompt[128];

	strncpy( prompt, str, 128 );
	for( ;; ){
		printf( "%s", prompt ); flush();
		if( oflag ) unixline( str );
		else jerqkbd( str );
		if( str[0] != '!' ) break;
		putchar( '\n' ); flush();
		system( &str[1] );
		printf( "\n!\n" );
	}
	return str;
}

char *doh( a )
{
	fmt( radix==16 ? "0x%X" : radix==10 ? "%u" : "0%o", a );
}

joff()
{
	int i;

	if( !oflag ){
		printf( "\r", STATELINE );
		jerqmenu( M_PENDULUM, 1 );
		flush();
		stabread( stabtab[MPX], 0 );
		jerqmenu( M_PENDULUM, 0 );
		hatebsmsg();
	} else newc();
	dotty = WORDDOT;
	radix = 10;
	dotfmt = DECIMAL;
	dotstrlen = 0;
	for( ;; ){
		tfree();
		scratch();
		if( !oflag ){
			if((hatebs = peeklong(HATEBS_INDEX)) == ACTIVE) train = 0;
			menuset();
		}
		sprintf( userline, dot ? "%s: " : ": ", doh(dot) );
		switch( *readuser( userline ) ){
		case ESCAPE:
			switch( userline[1] ){
			case DO_EVENT:
				hatebsmsg();
				break;
			case M_HIT:
				menuhit( (userline[2]<<8)|userline[3], userline[4]);
				break;
			}
			break;
		case '<' :
			putchar( '\n' );
			for( i = 1; userline[i] == ' '; ++i ){}
			stream(&userline[i]);
			break;
		case '\0' :
			putchar( oflag ? '\n' : '\v' );
			if( dot ){
				movedot( 1 );
				showdot();
			}
			break;
		default:
			if( !oflag ){
				m_menu( 0, 3 );
				m_menu( 0, 2 );
			}
			putchar( '\n' ); flush();
			lexindex = 0;
			strcat( userline, ";" );
			yyparse();
		}
	}
}

stream(script)
char *script;
{
	FILE *f;
	int i;

	if( !script[0] ) return;
	if( (f = fopen( script, "r" )) == NULL ){
		printf( "cannot open: %s\n", script );
		return;
	}
	while( fgets( userline, 256, f ) ){
		if( !strcmp( userline, "\n" ) ) continue;
		tfree();
		lexindex = 0;
		userline[strlen(userline)-1] = ';';
		printf( "%s \t ", userline );
		yyparse();
	}
	fclose( f );
}

tracebackc()
{
	if( hatebs == ACTIVE ) hatebsmsg();
	train = framechain( VERBOSE_CHAIN|FULL_CHAIN );
}

returnc( i )
{
	while( train && i-- ){
		if( train->caller == &sentinel ){
			printf( "use .t to obtain complete traceback\n" );
			return;
		}
		if( train->caller) train = train->caller;
	}
	context(train, LINE_CONTEXT|REGS_CONTEXT);
	putchar( '\n' );
}

callc( i )
{
	while( train && i-- ) if( train->callee) train = train->callee;
	context(train, LINE_CONTEXT|REGS_CONTEXT);
	putchar( '\n' );
}

framec()     { if( train ) dumpframe( train ); }

varsc()
{
	if( hatebs == ACTIVE ) hatebsmsg();
	if( train == 0) train = framechain( FULL_CHAIN );
	context( train, AUTO_CONTEXT|REGS_CONTEXT );
}

argvtables()
{
	char *str;
	struct ramnl *state, *fcn, *argv;
	MLONG argv0, udata, p = peeklong( PROC_INDEX );

	if( !(state = lookup( MPX, S_SSYM, "state", 0 ) ) ) return;
	if( !( peekword(p + state->n_value) & 8) ) return;	/* USER */

	if( !(fcn = lookup( MPX, S_SSYM, "fcn", 0 ) ) ) return;
	if( !(udata = peeklong( p + fcn->n_value ) ) ) return;

	if( !(argv = lookup( MPX, S_SSYM, "argv", 0 ) ) ) return;
	if( !(argv0 = peeklong( udata + argv->n_value ) ) )return;
	if( !(argv0 = peeklong( argv0 ) ) ) return;

	str = peeknstr( argv0, 128 );
	if( !str[0] ) return;
	if( strcmp( stabtab[USER]->path, str ) ){
		strcpy( stabtab[USER]->path, str );
		stabfree( stabtab[USER] );
	}
}

newc()
{
	int i;
	MLONG p = 1;

	train = 0;
	clrbpts();
	setbptscript(0);
	dot = snarfdot = 0;
	putchar( STATELINE );
	putchar( '\r' );
	if( !oflag ){
		flush();
		m_menu( 0, 2 ); 
		m_menu( 0, 3 ); 
		jerqdo( DO_DISABLE );
		jerqdo( DO_DEBUG );
		if( p = peeklong( PROC_INDEX) ){
			jerqdo( DO_VIDEO );
			argvtables();
		}
	}
	if( p ){
		stabread( stabtab[USER], oflag ? 0 : INTERACT );
		if( !oflag ){
			threadP(stabtab[USER]);
			makemap();
			hatebs = peeklong(HATEBS_INDEX);
			if( hatebs == ACTIVE || hatebs == ERRORED )
				hatebsmsg();
			jerqdo(DO_ENABLE);
		}
	}
}

inmemory( a )
MLONG a;
{
	if( a > LO_ADDR && a < HI_ADDR ) return 1;
	if( a >= LO_ROM && a < HI_ROM ) return 1;
	return 0;
}

dotop( i )
{
	switch( i ){
		default: assert(0);
		case 0:
		case MINUSONE:
		case PLUSONE:
			switch( i ){
			case MINUSONE: movedot( -1 ); break;
			case PLUSONE:  movedot( 1 ); break;
			}
			if( dotty == STRUCTDOT || dotty == STRINGDOT ) showdot();
			else printf( "%s: %s\n", doh(dot), fmtdot() );
			return;
		case INDIRECT: dot = peeklong(dot); return;
		case G_BITMAP:
		case G_POINT:
		case G_RECTANGLE:
		case G_TEXTURE:
			g_addr_desc( dot, i );
			return;
		case BYTEDOT: 
		case WORDDOT: 
		case LONGDOT: 
		case STRINGDOT: dotty = i; return;
		case STRUCTDOT: structenumdot( S_BSTR ); return;
		case ENUMDOT: structenumdot( S_BENUM ); return;
		case XREF: xrefc(); return;
		case DECIMAL: radix = 10; dotfmt = DECIMAL; return;
		case HEX: radix = 16; dotfmt = HEX; return;
		case OCTAL: radix = 8; dotfmt = OCTAL; return;
		case ASCII: dotfmt = ASCII; return;
		case SNARF: dot = snarfdot;
			    snarfdot = 0;
			    if( snarfdotty ) dotty = snarfdotty;
			    return;
	}
}

movedot( i )
{
	if( dotty != BYTEDOT && dotty != STRINGDOT )
		if( dot&1 ) ++dot;
	switch( dotty ){
		case BYTEDOT: dot += i; break;
		case WORDDOT: dot += i*2; break;
		case LONGDOT: dot += i*4; break;
		case STRINGDOT: dot += i*(dotstrlen ? dotstrlen : 4); break;
		case INSTDOT: dot += i == 1 ? delta : i; break;
		case ENUMDOT:
		case STRUCTDOT:  dot += i*delta; break;
		default: assert(0);
	}
	if( dot < 0 ) dot = 0;
}

showdot()
{
	int d, oh;

	if( dot == 0 ) return;
	if( dotty != BYTEDOT && dotty != STRINGDOT && dot&1 ) ++dot;
	switch( dotty ){
		case ENUMDOT:
		case STRUCTDOT: 
			dotstab[0].n_other = S_GSYM;
			dotstab[0].n_desc = (dotty == STRUCTDOT) ? STRTY : ENUMTY;
			dotstab[0].n_value = dot;
			printf( "%s: %s\n", doh(dot), display( &dotstab[0], 0 ) );
			return;
		case STRINGDOT:
			printf( "%s: %s\n", doh(dot), fmtstring( dot));
			return;
		case INSTDOT: delta = snap(dot) - dot; return;
		case BYTEDOT: oh = peekbyte( dot ); d = (int) (char) oh; break;
		case WORDDOT: oh = peekword( dot ); d = (int) (short) oh; break;
		case LONGDOT: d = oh = peeklong( dot ); break;
	}
	if( oh == 0 ){
		printf( "%s: 0\n", doh(dot) );
		return;
	}
	else printf( "%s: %d 0%o 0x%X \'", doh(dot), d , oh , oh );
	switch( dotty ){
		case LONGDOT:
			printf( "%s%s", fmtbyte( oh >> 24 ), fmtbyte( oh >> 16 ) );
		case WORDDOT: printf( "%s", fmtbyte( oh >> 8 ) );
		case BYTEDOT: printf( "%s", fmtbyte( oh ) );
	}
	printf( "\'\n" );
}

char *fmtshortstr( p )
MLONG p;
{
	int   width = 12, i = 0;
	char *buf = talloc( width+32 ), *s;

	if( !p ) return "0";
	assert( width >= 5 );
	strcpy( buf, "\"" );
	s = peeknstr( p, width );
	while( s[i] && strlen(buf) <= width ) strcat(buf, fmtbyte(s[i++]));
	strcat( buf, "\"" );
	if( strlen( buf ) > width ) strcpy( &buf[width-4], "...\"" );
	return buf;
}

char *fmtstruct( s )
struct ramnl *s;
{
	MLONG  p = locate( s, 0 );
	char *field, *val, *d;

#define	WIDTH 14
	(d = talloc(WIDTH+4))[0] = '{';
	if( (++s)->n_other != S_TYID ) return "<missing tables>";
	if( s->n_name[0] == '\0' ) return "<anon struct>";
	s = lookup( table(s), S_BSTR, s->n_name, 0 );
	if( !s ) return "<struct error>";
	for( ++s; s->n_other != S_ESTR; ){
		field = fmt( "%0.8s=", s->n_name );
		s->n_value += p;
		val =  display(s, 0);
		s->n_value -= p;
		if( strlen(field)+strlen(val)+strlen(d) >= WIDTH ) return d;
		sprintf( &d[strlen(d)], "%s%s", field, val );
		for( ++s; s->n_other == S_TYID || s->n_other == S_DIM; ++s ){}
		if( s->n_other == S_SSYM ) strcat( d, "," );
	}
	return strcat( d, "}" );
}

char *fmtdot()
{
	int d, oh, i;
	char *b[4];

	if( dot == 0 ) return "" ;
	if( dotty != BYTEDOT && dot&1 ) ++dot;
	switch( dotty ){
		case STRUCTDOT:
			dotstab[0].n_other = S_GSYM;
			dotstab[0].n_desc = STRTY;
			dotstab[0].n_value = dot;
			return fmtstruct( &dotstab[0] );
		case ENUMDOT:
			dotstab[0].n_other = S_GSYM;
			dotstab[0].n_desc = ENUMTY;
			dotstab[0].n_value = dot;
			return display( &dotstab[0], 0 );
		case STRINGDOT:
			return fmtshortstr( dot );
		case INSTDOT: assert(0);
		case BYTEDOT: oh = peekbyte( dot ); d = (int) (char) oh; break;
		case WORDDOT: oh = peekword( dot ); d = (int) (short) oh; break;
		case LONGDOT: d = oh = peeklong( dot ); break;
	}
	if( oh == 0 ) return "0";
	switch( dotfmt ){
	case DECIMAL: return fmt( "%d", d );
	case OCTAL:   return fmt( "0%o", oh );
	case HEX:     return fmt( "0x%X", oh );
	case ASCII:
		for( i = 0; i < 4; ++i ) b[i] = fmtbyte( oh >> (i*8) );
		switch( dotty ){
		case LONGDOT: return fmt( "\'%s%s%s%s\'", b[3], b[2], b[1], b[0] );
		case WORDDOT: return fmt( "\'%s%s\'", b[1], b[0] );
		case BYTEDOT: return fmt( "\'%s\'", b[0] );
		default: return "???";
		}
	}
}

idc( id )
char *id;
{
	struct ramnl *s, *sline;
	int glb = III( S_BFUN, S_GSYM, S_STSYM );
	int lcl = III(S_RSYM,S_LSYM,S_PSYM);;


	if( !train && hatebs != ACTIVE && hatebs != ERRORED ) train = framechain(0);
	if( ( train && (sline = lookup( USER_MPX, S_SLINE, 0, train->pc))
				&& (s = visible( sline, lcl, id, 0)) )
			|| ( s = lookup( USER_MPX, glb, id, 0)) ){}

	if( s ) printf( "%s\n", display(s,train));
	else if( (s = lookup( USER_MPX, LDE, id, 0) )
		|| (s = lookup( USER_MPX, LDL, id, 0)) ) 
			printf( "%s loaded at %s\n", id, doh(dot = s->n_value) );
	else if( s = lookup( USER_MPX, II(S_BSTR,S_BENUM), id, 0 ) ){
		dotty = s->n_other == S_BSTR ? STRUCTDOT : ENUMDOT;
		dotstab[1] = *s;
		dotstab[1].n_other = S_TYID;
		while( s->n_other != S_ESTR && s->n_other != S_EENUM ) ++s;
		delta = s->n_other == S_ESTR ? s->n_value : 2;
		showdot();
	}
	else printf( "not found: %s\n", id );
}

structenumdot( t )
long t;
{
	struct ramnl *s = lookup( USER, t, "*", 0 );
	long i;

	assert(s);
	scratch();
	for( ; s; s = s->n_thread ) scrsort( fmt( "%-8.8s", s->n_name ), s );
	if( ( i = scrhit(0) ) == -1 ) return;
	s = (struct ramnl *) i;
	dotty = s->n_other == S_BSTR ? STRUCTDOT : ENUMDOT;
	dotstab[1] = *s;
	dotstab[1].n_other = S_TYID;
	while( s->n_other != S_ESTR && s->n_other != S_EENUM ) ++s;
	delta = s->n_other == S_ESTR ? s->n_value : 2;
}
	
char *fmtxref()
{
	struct ramnl *s, *sline;
	MLONG  offset;
	struct frame *p;
	int r;

	for( p = train; p && p->caller && p->caller!= &sentinel; p = p->caller ){}
	for( ; p; p = p->callee )
	    if( p->sline )
		if( (s = visible( p->sline, S_PSYM, 0, dot - p->fp ) )
		 || (s = visible( p->sline, S_LSYM, 0, p->fp+p->regdelta-dot )))
			return fmt( "%0.8s in %s()", s->n_name, p->fname );
	if( s = lookup( USER_MPX, LDE, 0, dot ) ){
		if( (region(USER, dot) == N_TEXT || region(MPX, dot) == N_TEXT)
			&& (sline = lookup( USER_MPX, S_SLINE, 0, dot) )
				&& (offset = dot-sline->n_value) < 256 )
					return soline(dot);
		if( offset = dot - s->n_value )
			return fmt( "%0.8s+%s", s->n_name , doh(offset) );
		return fmt( "%0.8s", s->n_name );
	}
	return 0;
}
	
xrefc()
{
	struct ramnl *s;
	MLONG  offset;
	struct frame *p;
	int i, r;

	printf( "%s = ", doh( dot ) );
	for( i = 1; stabtab[i]; ++i ) if( r = region( i, dot ) )
		printf( "(%s %s) ", stabtab[i]->name, ntypenames[r] );
	for( p = train; p && p->caller && p->caller!= &sentinel; p = p->caller ){}
	for( ; p; p = p->callee )
	    if( p->sline )
		if( (s = visible( p->sline, S_PSYM, 0, dot - p->fp ) )
		 || (s = visible( p->sline, S_LSYM, 0, p->fp+p->regdelta-dot ))){
			printf( "%0.8s in %s()\n", s->n_name, p->fname );
			return;
		}
	if( s = lookup( USER_MPX, LDE, 0, dot ) ){
		printf( "%0.8s", s->n_name );
		if( offset = dot - s->n_value ) printf( "+%s", doh(offset) );
		if( (region(USER, dot) == N_TEXT || region(MPX, dot) == N_TEXT)
			&& (s = lookup( USER_MPX, S_SLINE, 0, dot) )
				&& (offset = dot-s->n_value) < 256 ){
			printf( " %s", soline(dot) );
		 	printf(" in %0.8s()",
					lookup(table(s),S_BFUN,0,dot)->n_name);
		}
		putchar( '\n' );
		return;
	}
	printf( "no symbolic form\n" );
}

regc( r )
{
	struct ramnl reg, *s;

	if( !train ){
		if( hatebs == STEPPED ) tracereg( r, peeklong(TLOC_INDEX) );
		else printf( "<no context>\n" );
		return;
	}
	if( r == -1  ){
		for( r = 0; r <= 15; ++r ) regc( r );
		return;
	}
	printf( "%s ", regnames[r] );
	if( s = visible( train->sline, S_RSYM, 0, r))
				printf( "(%0.8s) ", s->n_name );
	reg.n_other = S_RSYM;
	reg.n_value = r;
	if( dot = locate( &reg, train ) ){
		dotty = LONGDOT;
		showdot();
	}
	else printf( "not saved\n" );
}

tracereg( r, tloc )
MLONG tloc;
{
	switch( r ){
	case -1:
		for( r = 0; r <= 15; ++r ) tracereg( r, tloc );
		break;
	case 15:
		printf( "a7 (%%sp): %s\n", doh(tloc) );
		break;
	default:
		dot = tloc-2-14*4+( r==14 ? -3 : r )*4;	/* fp saved by trap() */
		dotty = LONGDOT;
		printf( "%s ", regnames[r] );
		showdot();
		return;
	}
}

struct ramnl *fileline( file, line )
char *file;
{
	struct ramnl *so = lookup( USER, S_SO, "*", 0 ), *sline, *nso;
	char *sof;

	if( !so ){
		printf( "missing tables\n" );
		return 0;
	}
	for( ; so; so = so->n_thread ){
		if( !strcmp( file, sof = sofile(so) ) ) break;
		if( !strcmp( file, basename(sof) ) ) break;
	}
	if( !so ){
		printf( "not found: %s\n", file );
		return 0;
	}
	for( sline = so; sline->n_other != S_SLINE; ++sline ){
		if( sline->n_other == S_EFUN ){
			printf( "missing tables\n" );
			return 0;
		}
	}
	for( nso = so; nso->n_thread == nso+1; ++nso ) {}
	nso = nso->n_thread;
	for( ; sline; sline = sline->n_thread ){
		if( nso && sline > nso ) break;
		if( sline->n_desc >= line ){
			if( sline->n_desc != line )
			printf( "using %d instead of %d\n", sline->n_desc, line );
			return sline;
		}
	}
	printf( "line %d not found in %s\n", line, sofile(so) );
	return 0;
}
	
srclinec( file, line )
char *file;
{
	struct ramnl *sline;

	if( !(sline = fileline( file, line ))) return;
	dot = sline->n_value;
	dotty = INSTDOT;
	xrefc();
	showdot();
} 	

quitc()
{
	if( !oflag && !menuconfirm() ) return;
	clean = 1;
	wrap();
}

zmc(f)
char *f;
{
	stabfree( stabtab[MPX] );
	strcpy( stabtab[MPX]->path, f );
	stabread( stabtab[MPX], 0);
}
